src/modules/job-exec/exec.c

```diff
@@ -227,21 +236,28 @@ static void error_cb (struct bulk_exec *exec, flux_subprocess_t *p, void *arg)
      */
     if (cmd) {
         if (errnum == EHOSTUNREACH) {
-            if (!idset_test (job->critical_ranks, shell_rank)
-                && lost_shell (job,
-                               true,
-                               shell_rank,
-                               "%s on %s (shell rank %d)",
-                               "lost contact with job shell",
-                               hostname,
-                               shell_rank) == 0)
-                return;
-            jobinfo_fatal_error (job,
-                                0,
-                                "%s on broker %s (rank %d)",
-                                "lost contact with job shell",
-                                hostname,
-                                rank);
+            bool critical = is_critical_rank (job, shell_rank);
+
+            /*  Always notify rank 0 shell of a lost shell.
+             */
+            lost_shell (job,
+                        critical,
+                        shell_rank,
+                        "%s on %s (shell rank %d)",
+                        "lost contact with job shell",
+                        hostname,
+                        shell_rank);
+
+            /*  Raise a fatal error and terminate job immediately if
+             *  the lost shell was critical.
+             */
+            if (critical)
+                jobinfo_fatal_error (job,
+                                     0,
+                                     "%s on broker %s (rank %d)",
+                                     "lost contact with job shell",
+                                     hostname,
+                                     rank);
```

Comment:
would it be useful to say "critical" somewhere in this message to differentiate it from the prior one?  Like "lost contact with critical job shell"?  I'm aware the messages are going to different places ... but might be useful when debugging?

Reply:
Perhaps, though it is probably enough that the exception raised will be fatal.

Note that by far the common case is that all job shells are critical. Only in the case of a Flux instance (which adjusts its own critical ranks to remove the leaf broker rank), or if a user manually updates critical ranks will the non critical (severity 1) `lost-shell` exception be raised.

Because this is the common case, it might be less confusing to keep the error message to the user simpler, at least for now.

Reply:
Also, in case it wasn't obvious from the diff, this particular call didn't change from before, the code was just reorganized to also call `lost_shell()` for a critical rank loss.